### PR TITLE
Integrate current execution check with replication resender

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
1. Integrate current execution check with replication resender
2. Update the current execution check to make sure current execution exists. 

**Why?**
Reduce the number of messages go into DLQ

**How did you test it?**
Unit tests and local tests

**Potential risks**
N/A

### CODE REVIEW

File: common/reconciliation/invariants/concreteExecutionExists.go [hunk 1]
```diff
@@ -59,6 +60,11 @@ func (c *concreteExecutionExists) Check(execution interface{}) common.CheckResul
 			InvariantType:   c.InvariantType(),
 		}
 	}
```
Comment (yycptt) [inline common/reconciliation/invariants/concreteExecutionExists.go:62]: Not related to your change but is the check from L57-62 correct? Even if the state for current execution is closed, as long as corresponding concrete execution doesn't exist, it's still a corruption I think?

File: common/reconciliation/invariants/concreteExecutionExists.go [hunk 2]
```diff
@@ -59,6 +60,11 @@ func (c *concreteExecutionExists) Check(execution interface{}) common.CheckResul
...
+	var runIDCheckResult *common.CheckResult
+	currentExecution, runIDCheckResult = c.validateCurrentRunID(currentExecution)
+	if runIDCheckResult != nil {
+		return *runIDCheckResult
+	}
 	// by this point the corresponding concrete execution must exist and can be already closed
```
Comment (yycptt) [inline common/reconciliation/invariants/concreteExecutionExists.go:68]: Can we remove this comment? I think it's kind of misleading. If we are sure that corresponding concrete execution must exist here, then we don't need the `IsWorkflowExecutionExists` call below and can return `CheckResultTypeHealthy` directly I think?

File: common/reconciliation/invariants/concreteExecutionExists.go [hunk 3]
```diff
@@ -59,6 +60,11 @@ func (c *concreteExecutionExists) Check(execution interface{}) common.CheckResul
...
 	currentExecResp, currentExecErr := c.pr.IsWorkflowExecutionExists(&persistence.IsWorkflowExecutionExistsRequest{
```
Comment (yycptt) [inline common/reconciliation/invariants/concreteExecutionExists.go:69]: Also not related to your change, but I think there's a race condition here. The workflow retention period can kick in before we make this call, so even if the response says the concrete execution doesn't exist, we can't be 100% sure it's a corruption, we need to check if the current execution with the same runID is still there.

another nit: can we rename the var from `currentExecResp, currentExecErr` to `concreteExceResp, concreteExecErr`?

File: common/reconciliation/invariants/concreteExecutionExists.go [hunk 4]
```diff
@@ -89,11 +95,20 @@ func (c *concreteExecutionExists) Check(execution interface{}) common.CheckResul
 }
 
 func (c *concreteExecutionExists) Fix(execution interface{}) common.FixResult {
-	fixResult, checkResult := checkBeforeFix(c, execution)
+	currentExecution, _ := execution.(*common.CurrentExecution)
+	var runIDCheckResult *common.CheckResult
+	currentExecution, runIDCheckResult = c.validateCurrentRunID(currentExecution)
```
Comment (yycptt) [inline common/reconciliation/invariants/concreteExecutionExists.go:100]: Is this needed? The deletion will be a no-op when `runIDCheckResult != nil` I think?

File: common/reconciliation/invariants/concreteExecutionExists.go (hunk 1 above)
Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go:62]: Yes. You are correct

File: common/reconciliation/invariants/concreteExecutionExists.go (hunk 2 above)
Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go:68]: Removed

File: common/reconciliation/invariants/concreteExecutionExists.go (hunk 3 above)
Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go:69]: Yes. You are right. The deletion of the current execution will check the current run ID.

I will update the namings.

File: common/reconciliation/invariants/concreteExecutionExists.go (hunk 4 above)
Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go:100]: Yes. For scanner, the current run ID is set in the execution. For resender, it passes in the domain and workflow ID. This is to set the current run ID in this case.

File: common/reconciliation/invariants/concreteExecutionExists.go (hunk 3 above)
Comment (mkolodezny) [reply common/reconciliation/invariants/concreteExecutionExists.go]: the retention kicks in after the workflow gets closed only but there is a check for the open state above.

Comment (yycptt) [reply common/reconciliation/invariants/concreteExecutionExists.go]: Synced offline. We will also cover the case where state is closed.

File: common/reconciliation/invariants/concreteExecutionExists.go (hunk 4 above)
Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go]: Updated the code after offline sync

File: common/reconciliation/invariants/concreteExecutionExists.go [hunk 5]
```diff
@@ -115,3 +142,44 @@ func (c *concreteExecutionExists) Fix(execution interface{}) common.FixResult {
...
+		}
+	}
+
+	if len(currentExecution.CurrentRunID) == 0 {
+		currentExecution.CurrentRunID = resp.RunID
+	}
+
+	if currentExecution.CurrentRunID != resp.RunID {
+		return nil, &common.CheckResult{
+			CheckResultType: common.CheckResultTypeFailed,
```
Comment (yycptt) [inline common/reconciliation/invariants/concreteExecutionExists.go:179]: actual I think we can return healthy here? Seems like it's the same thing as current record with this runID doesn't exist.

Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go]: Yes. Updated

File: common/reconciliation/invariants/concreteExecutionExists.go [hunk 6]
```diff
@@ -53,27 +57,35 @@ func (c *concreteExecutionExists) Check(execution interface{}) common.CheckResul
...
-			InfoDetails:     currentExecErr.Error(),
+			InfoDetails:     concreteExecErr.Error(),
 		}
 	}
-	if !currentExecResp.Exists {
+	if !concreteExecResp.Exists {
+		//verify if the current execution exists
+		_, checkResult := c.validateCurrentRunID(currentExecution)
+		if checkResult != nil {
+			return *checkResult
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/reconciliation/invariants/concreteExecutionExists.go:87]: Updated in line 179

**Pull request merged**